src/IO/S3/deleteFileFromS3.cpp

```diff
@@ -0,0 +1,235 @@
+#include <IO/S3/deleteFileFromS3.h>
+
+#if USE_AWS_S3
+
+#include <Common/logger_useful.h>
+#include <IO/S3/Client.h>
+#include <IO/S3/Requests.h>
+#include <IO/S3/BlobStorageLogWriter.h>
+#include <IO/S3/S3Capabilities.h>
+#include <IO/S3/getObjectInfo.h>
+
+
+namespace ProfileEvents
+{
+    extern const Event S3DeleteObjects;
+}
+
+
+namespace DB
+{
+
+void deleteFileFromS3(
+    const std::shared_ptr<const S3::Client> & s3_client,
+    const String & bucket,
+    const String & key,
+    bool if_exists,
+    BlobStorageLogWriterPtr blob_storage_log,
+    const String & local_path_for_blob_storage_log,
+    size_t file_size_for_blob_storage_log,
+    std::optional<ProfileEvents::Event> profile_event)
+{
+    ProfileEvents::increment(ProfileEvents::S3DeleteObjects);
+    if (profile_event && *profile_event != ProfileEvents::S3DeleteObjects)
+        ProfileEvents::increment(*profile_event);
+
+    auto log = getLogger("deleteFileFromS3");
+
+    S3::DeleteObjectRequest request;
+    request.SetBucket(bucket);
+    request.SetKey(key);
+    auto outcome = s3_client->DeleteObject(request);
+
+    if (blob_storage_log)
+    {
+        blob_storage_log->addEvent(BlobStorageLogElement::EventType::Delete,
+                                   bucket, key,
+                                   local_path_for_blob_storage_log, file_size_for_blob_storage_log,
+                                   outcome.IsSuccess() ? nullptr : &outcome.GetError());
+    }
+    else
+    {
+        LOG_TRACE(log, "No blob storage log, not writing blob {}", key);
+    }
+
+    if (!outcome.IsSuccess() && (!if_exists || !S3::isNotFoundError(outcome.GetError().GetErrorType())))
```

Comment:
I do not really understand why we do not rethrow on the error RESOURCE_NOT_FOUND but rethrow on the error NO_SUCH_BUCKET, but I do not have any particular suggestion, jfyi

Reply:
I didn't change `S3::isNotFoundError()` before, but I suppose there is no harm in checking NO_SUCH_BUCKET too. [Done](https://github.com/ClickHouse/ClickHouse/pull/70786/files#diff-b1ee988246d11cd3c6981d6b4b2a894a178f74ed9467c4a76aaa833023e71f75R69-R70).